File: feral-nodes/ts-node-sdk/src/node.ts
/*
 * FeralNode — the public class every TypeScript/Node.js hardware daemon is built on.
 * Mirrors the Python `FeralNode` surface: auto-reconnect with jittered backoff,
 * automatic heartbeat loop, outbound schema validation (HUP v1, see ../../HUP_SPEC.md),
 * optional mDNS discovery and 6-digit pairing helpers.
 */

import WebSocket from "ws";
import { Capability, CapabilityName } from "./capability";
import { discoverBrain as _discoverBrain } from "./discovery";
import { loadKey as _loadKey, pair as _pair, PairOptions } from "./pairing";
import {
  buildFrame,
  DeviceEventPayload,
  HUP_VERSION,
  HUPActionRequestPayload,
  HUPActionResponsePayload,
  NodeAckPayload,
  NodeByePayload,
  NodeHeartbeatPayload,
  NodeRegisterPayload,
  NodeType,
} from "./schemas";

export type ActionHandler = (
  params: Record<string, unknown>,
) => Promise<Record<string, unknown> | void> | Record<string, unknown> | void;

export interface FeralNodeOptions {
  nodeId: string;
  name?: string;
  firmwareVersion?: string;
  brainUrl?: string;
  apiKey?: string;
  capabilities?: Array<CapabilityName | string>;
  nodeType?: NodeType;
  manufacturer?: string;
  model?: string;
  platform?: string;
  sensors?: string[];
  actuators?: string[];
  location?: string;
  tags?: string[];
  heartbeatMs?: number;
}

const PASSIVE_SENSOR_CAPS = new Set<string>([
  Capability.HEART_RATE,
  Capability.SPO2,
  Capability.TEMPERATURE,
  Capability.UV,
  Capability.ACCELEROMETER,
  Capability.GYROSCOPE,
  Capability.AMBIENT_LIGHT,
  Capability.STEPS,
  Capability.BATTERY,
  Capability.GPS,
  Capability.MICROPHONE,
  Capability.CAMERA,
]);

const ACTUATOR_CAPS = new Set<string>([
  Capability.DISPLAY,
  Capability.SPEAKER,
  Capability.HAPTIC,
  Capability.BUZZER,
  Capability.LED,
  Capability.MOTOR,
  Capability.RELAY,
  Capability.VALVE,
]);

export class FeralNode {
  readonly nodeId: string;
  readonly name: string;
  readonly firmwareVersion: string;
  readonly nodeType: NodeType;
  readonly capabilities: string[];
  readonly sensors: string[];
  readonly actuators: string[];

  private brainUrl: string | undefined;
  private apiKeyOverride: string | undefined;
  private manufacturer: string;
  private model: string;
  private platform: string;
  private location: string;
  private tags: string[];
  private heartbeatMs: number;

  private ws: WebSocket | null = null;
  private handlers = new Map<string, ActionHandler>();
  private stopping = false;
  private granted = new Set<string>();
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  private sessionToken: string | null = null;
  private connectedResolver: (() => void) | null = null;
  private connectedPromise: Promise<void>;

  constructor(opts: FeralNodeOptions) {
    this.nodeId = opts.nodeId;
    this.name = opts.name ?? opts.nodeId;
    this.firmwareVersion = opts.firmwareVersion ?? "0.0.0";
    this.brainUrl = opts.brainUrl;
    this.apiKeyOverride = opts.apiKey;
    this.nodeType = (opts.nodeType ?? "sensor") as NodeType;
    this.manufacturer = opts.manufacturer ?? "";
    this.model = opts.model ?? "";
    this.platform = opts.platform ?? "";
    this.location = opts.location ?? "";
    this.tags = opts.tags ?? [];
    this.heartbeatMs = opts.heartbeatMs ?? 10_000;

    this.capabilities = (opts.capabilities ?? []).map(String);
    this.sensors = opts.sensors ?? this.capabilities.filter((c) => PASSIVE_SENSOR_CAPS.has(c));
    this.actuators = opts.actuators ?? this.capabilities.filter((c) => ACTUATOR_CAPS.has(c));

    this.connectedPromise = new Promise<void>((res) => {
      this.connectedResolver = res;
    });
  }

  onAction(name: string, handler: ActionHandler): this {
    this.handlers.set(name, handler);
    return this;
  }

  async emitEvent(
    eventType: string,
    data: Record<string, unknown>,
  ): Promise<void> {
    const payload: DeviceEventPayload = {
      node_id: this.nodeId,
      event_type: eventType,
      data,
      ts: Date.now() / 1000,
    };
    await this.send("device_event", payload);
  }

  static async discoverBrain(timeoutMs = 3000): Promise<string | null> {
    return _discoverBrain(timeoutMs);
  }

  static async pair(opts: PairOptions): Promise<string> {
    return _pair(opts);
  }

  async run(mainLoop?: () => Promise<void>): Promise<void> {
    const supervisor = this.wsSupervisor();
    if (!mainLoop) {
      await supervisor;
      return;
    }
    const main = (async () => {
      await this.connectedPromise;
      await mainLoop();
    })();
    await Promise.race([supervisor, main]);
    this.stopping = true;
  }

  async close(reason = "shutdown"): Promise<void> {
    this.stopping = true;
    try {
      await this.send("node_bye", { reason, restart_in_s: 0 } as NodeByePayload);
    } catch {
      /* ignore */
    }
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
    if (this.ws) {
      try {
        this.ws.close();
      } catch {
        /* ignore */
      }
      this.ws = null;
    }
  }

  private async resolveApiKey(): Promise<string | null> {
    if (this.apiKeyOverride) return this.apiKeyOverride;
    return _loadKey(this.nodeId);
  }

  private async wsSupervisor(): Promise<void> {
    let backoff = 100;
    while (!this.stopping) {
      const url = this.brainUrl ?? (await _discoverBrain(3000));
      if (!url) {
        await this.sleep(backoff);
        backoff = Math.min(backoff * 2, 30_000);
        continue;
      }

      const apiKey = await this.resolveApiKey();
      const headers: Record<string, string> = apiKey
        ? { Authorization: `Bearer ${apiKey}` }
        : {};

      try {
        await this.runOnce(url, headers);
        backoff = 100;
      } catch (err) {
        // eslint-disable-next-line no-console
        console.warn(`[feral-node-sdk] ws error: ${(err as Error).message}`);
      } finally {
        this.ws = null;
        if (this.heartbeatTimer) {
          clearInterval(this.heartbeatTimer);
          this.heartbeatTimer = null;
        }
      }

      if (this.stopping) return;
      const jitter = backoff * (0.5 + Math.random());
      await this.sleep(Math.min(jitter, 30_000));
      backoff = Math.min(backoff * 2, 30_000);
    }
  }

  private runOnce(
    url: string,
    headers: Record<string, string>,
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(url, { headers, maxPayload: 2 * 1024 * 1024 });
      this.ws = ws;

      ws.on("open", () => {
        this.handshake().catch(reject);
      });

      ws.on("message", (raw) => {
        this.handleFrame(raw.toString()).catch((err) => {
          // eslint-disable-next-line no-console
          console.warn("[feral-node-sdk] frame handler error:", err);
        });
      });

      ws.on("error", (err) => reject(err));
      ws.on("close", () => resolve());
    });
  }

  private async handshake(): Promise<void> {
    const payload: NodeRegisterPayload = {
      node_id: this.nodeId,
      node_type: this.nodeType,
      name: this.name,
      manufacturer: this.manufacturer,
      model: this.model,
      firmware_version: this.firmwareVersion,
      platform: this.platform,
      os: "",
      capabilities: this.capabilities,
      sensors: this.sensors,
      actuators: this.actuators,
      location: this.location,
      tags: this.tags,
    };
    await this.send("node_register", payload);
  }

  private async handleFrame(raw: string): Promise<void> {
    let frame: { type?: string; payload?: Record<string, unknown> };
    try {
      frame = JSON.parse(raw);
    } catch {
      return;
    }
    const type = frame.type ?? "";
    const payload = (frame.payload ?? {}) as Record<string, unknown>;

    switch (type) {
      case "node_ack": {
        const ack = payload as unknown as NodeAckPayload;
        this.sessionToken = ack.session_token ?? null;
        this.granted = new Set(
          (ack.granted_capabilities && ack.granted_capabilities.length)
            ? ack.granted_capabilities
            : this.capabilities,
        );
        this.heartbeatMs = Math.max(1000, Number(ack.heartbeat_ms) || this.heartbeatMs);
        this.startHeartbeat();
        this.connectedResolver?.();
        return;
      }
      case "hup_action_request": {
        const req = payload as unknown as HUPActionRequestPayload;
        await this.dispatchAction(req);
        return;
      }
      default:
        return;
    }
  }

  private async dispatchAction(req: HUPActionRequestPayload): Promise<void> {
    const handler = this.handlers.get(req.name);
    if (!handler) {
      await this.sendActionResponse(req.action_id, {
        success: false,
        error: `capability_denied: ${req.name}`,
      });
      return;
    }
    const started = Date.now();
    try {
      const result = await handler(req.params ?? {});
      await this.sendActionResponse(req.action_id, {
        success: true,
        result: (result && typeof result === "object" ? (result as Record<string, unknown>) : { ok: true }),
        duration_ms: Date.now() - started,
      });
    } catch (err) {
      await this.sendActionResponse(req.action_id, {
        success: false,
        error: (err as Error).message,
        duration_ms: Date.now() - started,
      });
    }
  }

  private async sendActionResponse(
    actionId: string,
    extras: Partial<HUPActionResponsePayload> & { success: boolean },
  ): Promise<void> {
    const payload: HUPActionResponsePayload = {
      action_id: actionId,
      success: extras.success,
      result: extras.result ?? {},
      error: extras.error ?? null,
      duration_ms: extras.duration_ms ?? 0,
    };
    await this.send("hup_action_response", payload);
  }

  private startHeartbeat(): void {
    if (this.heartbeatTimer) clearInterval(this.heartbeatTimer);
    this.heartbeatTimer = setInterval(() => {
      const payload: NodeHeartbeatPayload = { ts: Date.now() / 1000 };
      this.send("node_heartbeat", payload).catch(() => undefined);
    }, this.heartbeatMs);
  }

  private async send(type: string, payload: Record<string, unknown>): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    const frame = buildFrame(type, payload);
    await new Promise<void>((resolve, reject) => {
      this.ws!.send(JSON.stringify(frame), (err) => (err ? reject(err) : resolve()));
    });
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((r) => setTimeout(r, ms));
  }
}

export { HUP_VERSION, Capability };
export type { CapabilityName, PairOptions };

File: feral-nodes/python-node-sdk/src/feral_node_sdk/node.py
"""`FeralNode` — the public class every Python hardware daemon is built on.

Provides a clean async surface (`on_action`, `emit_event`, `run`) over the
HUP v1 wire protocol defined in `HUP_SPEC.md`: single persistent WebSocket,
auto-reconnect with jittered exponential backoff, automatic heartbeat loop,
outbound schema validation, optional mDNS discovery, and first-time pairing.
"""

from __future__ import annotations

import asyncio
import json
import logging
import random
import time
from typing import Any, Awaitable, Callable, Iterable, Optional, Union

import websockets
from websockets.client import WebSocketClientProtocol

from . import pairing as _pairing
from .capability import Capability
from .discovery import discover_brain as _discover_brain
from .schemas import (
    HUP_VERSION,
    DeviceEventPayload,
    HUPActionResponsePayload,
    MESSAGE_TYPES,
    NodeByePayload,
    NodeHeartbeatPayload,
    NodeRegisterPayload,
    build_frame,
)

logger = logging.getLogger("feral_node_sdk")

ActionHandler = Callable[[dict[str, Any]], Awaitable[Any]]
CapLike = Union[Capability, str]


class FeralNode:
    """High-level HUP v1 node client.

    Typical lifecycle:

        node = FeralNode(node_id=..., capabilities=[...], ...)

        @node.on_action("buzz")
        async def buzz(params): ...

        node.run(main_coro())

    The `run()` call blocks until `main_coro()` completes or Ctrl-C.
    """

    def __init__(
        self,
        *,
        node_id: str,
        name: str = "",
        firmware_version: str = "0.0.0",
        brain_url: Optional[str] = None,
        api_key: Optional[str] = None,
        capabilities: Iterable[CapLike] = (),
        node_type: str = "sensor",
        manufacturer: str = "",
        model: str = "",
        platform: str = "",
        sensors: Optional[Iterable[str]] = None,
        actuators: Optional[Iterable[str]] = None,
        location: str = "",
        tags: Optional[Iterable[str]] = None,
        heartbeat_ms: int = 10_000,
        log_level: int = logging.INFO,
    ) -> None:
        logging.basicConfig(level=log_level, format="%(asctime)s [%(levelname)s] %(name)s: %(message)s")
        self.node_id = node_id
        self.name = name or node_id
        self.firmware_version = firmware_version
        self.brain_url = brain_url
        self._api_key_override = api_key
        self.node_type = node_type
        self.manufacturer = manufacturer
        self.model = model
        self.platform = platform
        self.location = location
        self.tags: list[str] = list(tags or [])

        caps: list[str] = [c.value if isinstance(c, Capability) else str(c) for c in capabilities]
        self.capabilities: list[str] = caps
        self.sensors: list[str] = list(sensors) if sensors is not None else [
            c for c in caps if c in {
                "heart_rate","spo2","temperature","uv","accelerometer","gyroscope",
                "ambient_light","steps","battery","gps","microphone","camera",
            }
        ]
        self.actuators: list[str] = list(actuators) if actuators is not None else [
            c for c in caps if c in {
                "display","speaker","haptic","buzzer","led","motor","relay","valve",
            }
        ]

        self._heartbeat_ms = heartbeat_ms
        self._action_handlers: dict[str, ActionHandler] = {}
        self._ws: Optional[WebSocketClientProtocol] = None
        self._send_lock = asyncio.Lock()
        self._stop = asyncio.Event()
        self._granted: set[str] = set()
        self._session_token: Optional[str] = None
        self._connected = asyncio.Event()

    # ───────────────────────── public API ─────────────────────────

    def on_action(self, name: str) -> Callable[[ActionHandler], ActionHandler]:
        """Decorator: register an async handler for an `hup_action_request.name`."""

        def _wrap(fn: ActionHandler) -> ActionHandler:
            self._action_handlers[name] = fn
            return fn

        return _wrap

    async def emit_event(self, event_type: str, data: dict[str, Any]) -> None:
        """Send a validated `device_event` frame to the brain."""
        payload = DeviceEventPayload(
            node_id=self.node_id,
            event_type=event_type,
            data=data,
            ts=time.time(),
        )
        await self._send("device_event", payload)

    async def emit_audio_frame(
        self,
        data_b64: str,
        *,
        codec: str = "opus",
        sample_rate: int = 24000,
        channels: int = 1,
        sequence: int = 0,
        frame_ms: int = 20,
    ) -> None:
        """Send a HUP v1.1 ``audio_frame`` to the brain.

        The SDK validates the payload against ``AudioFramePayload`` before
        sending, so an oversize or malformed frame raises locally instead
        of being rejected by the brain with error 4020.
        """
        from .schemas import AudioFramePayload, DeviceEventPayload as _DE

        af = AudioFramePayload(
            codec=codec,
            sample_rate=sample_rate,
            channels=channels,
            frame_ms=frame_ms,
            sequence=sequence,
            data_b64=data_b64,
        )
        envelope = _DE(
            node_id=self.node_id,
            event_type="audio_frame",
            data=af.model_dump(),
            ts=time.time(),
        )
        await self._send("device_event", envelope)

    async def emit_video_frame(
        self,
        data_b64: str,
        *,
        codec: str = "jpeg",
        width: int,
        height: int,
        sequence: int = 0,
        keyframe: bool = True,
    ) -> None:
        """Send a HUP v1.1 ``video_frame`` to the brain."""
        from .schemas import VideoFramePayload, DeviceEventPayload as _DE

        vf = VideoFramePayload(
            codec=codec,
            width=width,
            height=height,
            sequence=sequence,
            keyframe=keyframe,
            data_b64=data_b64,
        )
        envelope = _DE(
            node_id=self.node_id,
            event_type="video_frame",
            data=vf.model_dump(),
            ts=time.time(),
        )
        await self._send("device_event", envelope)

    @staticmethod
    async def discover_brain(timeout_s: float = 3.0) -> Optional[str]:
        """Resolve a FERAL brain on the LAN via mDNS. See `discovery.py`."""
        return await _discover_brain(timeout_s=timeout_s)

    @classmethod
    async def pair(
        cls,
        node_id: str,
        brain_url: str,
        *,
        code: Optional[str] = None,
        name: str = "",
        timeout_s: float = 300.0,
        verify_tls: bool = True,
    ) -> str:
        """Run the 6-digit pairing flow. Returns the persisted API key."""
        return await _pairing.pair(
            node_id=node_id,
            brain_url=brain_url,
            name=name,
            code=code,
            timeout_s=timeout_s,
            verify_tls=verify_tls,
        )

    def run(self, main_coro: Optional[Awaitable[Any]] = None) -> None:
        """Block: connect, run `main_coro` (if given), keep reconnecting forever."""
        try:
            asyncio.run(self._run_async(main_coro))
        except KeyboardInterrupt:
            logger.info("Daemon stopped by user.")

    async def close(self, reason: str = "shutdown") -> None:
        """Send `node_bye` and tear down the current connection."""
        self._stop.set()
        try:
            await self._send("node_bye", NodeByePayload(reason=reason))
        except Exception:
            pass
        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:
                pass

    # ─────────────────────── internal machinery ───────────────────

    def _resolve_api_key(self) -> Optional[str]:
        return self._api_key_override or _pairing.load_key(self.node_id)

    async def _run_async(self, main_coro: Optional[Awaitable[Any]]) -> None:
        main_task: Optional[asyncio.Task[Any]] = None
        try:
            ws_task = asyncio.create_task(self._ws_supervisor(), name="feral-ws")
            if main_coro is not None:
                main_task = asyncio.create_task(self._run_main(main_coro), name="feral-main")
                done, _ = await asyncio.wait({ws_task, main_task}, return_when=asyncio.FIRST_COMPLETED)
                for t in done:
                    t.result()
            else:
                await ws_task
        finally:
            self._stop.set()
            if main_task and not main_task.done():
                main_task.cancel()
            await self.close()

    async def _run_main(self, coro: Awaitable[Any]) -> None:
        await self._connected.wait()
        await coro

    async def _ws_supervisor(self) -> None:
        backoff = 0.1
        while not self._stop.is_set():
            url = self.brain_url or await _discover_brain(timeout_s=3.0)
            if not url:
                logger.warning("No brain_url set and mDNS found none. Retrying in %.1fs", backoff)
                await asyncio.sleep(backoff)
                backoff = min(backoff * 2, 30.0)
                continue

            key = self._resolve_api_key()
            headers = [("Authorization", f"Bearer {key}")] if key else []

            try:
                logger.info("Connecting to %s", url)
                async with websockets.connect(
                    url,
                    additional_headers=headers,
                    max_size=2 * 1024 * 1024,
                    ping_interval=None,
                ) as ws:
                    self._ws = ws
                    backoff = 0.1
                    await self._handshake()
                    self._connected.set()
                    await asyncio.gather(
                        self._read_loop(),
                        self._heartbeat_loop(),
                    )
            except Exception as exc:
                logger.warning("WS connection error: %s", exc)
            finally:
                self._ws = None
                self._connected.clear()

            if self._stop.is_set():
                break

            sleep_for = backoff * (0.5 + random.random())
            sleep_for = min(sleep_for, 30.0)
            logger.info("Reconnecting in %.2fs", sleep_for)
            await asyncio.sleep(sleep_for)
            backoff = min(backoff * 2, 30.0)

    async def _handshake(self) -> None:
        payload = NodeRegisterPayload(
            node_id=self.node_id,
            node_type=self.node_type,  # type: ignore[arg-type]
            name=self.name,
            manufacturer=self.manufacturer,
            model=self.model,
            firmware_version=self.firmware_version,
            platform=self.platform,
            capabilities=self.capabilities,
            sensors=self.sensors,
            actuators=self.actuators,
            location=self.location,
            tags=self.tags,
        )
        await self._send("node_register", payload)

    async def _read_loop(self) -> None:
        assert self._ws is not None
        async for raw in self._ws:
            try:
                frame = json.loads(raw)
            except Exception:
                logger.warning("Dropped non-JSON frame")
                continue
            msg_type = frame.get("type", "")
            payload = frame.get("payload") or {}
            if msg_type == "node_ack":
                self._session_token = payload.get("session_token")
                self._granted = set(payload.get("granted_capabilities") or self.capabilities)
                hb = int(payload.get("heartbeat_ms") or self._heartbeat_ms)
                self._heartbeat_ms = max(1000, hb)
                logger.info("node_ack: granted=%s heartbeat=%dms",
                            sorted(self._granted), self._heartbeat_ms)
            elif msg_type == "hup_action_request":
                asyncio.create_task(self._dispatch_action(payload))
            elif msg_type in ("text_response", "error", "node_bye"):
                logger.debug("frame %s: %s", msg_type, payload)
            else:
                logger.debug("ignoring unknown frame type: %s", msg_type)

    async def _dispatch_action(self, payload: dict[str, Any]) -> None:
        action_id = str(payload.get("action_id", ""))
        name = str(payload.get("name", ""))
        params = dict(payload.get("params") or {})

        handler = self._action_handlers.get(name)
        if handler is None:
            await self._send_action_response(
                action_id, success=False, error=f"capability_denied: {name}"
            )
            return

        started = time.time()
        try:
            result = await handler(params)
        except Exception as exc:
            logger.exception("action %s failed", name)
            await self._send_action_response(
                action_id, success=False, error=str(exc),
                duration_ms=int((time.time() - started) * 1000),
            )
            return

        if not isinstance(result, dict):
            result = {"ok": True, "value": result}
        await self._send_action_response(
            action_id, success=True, result=result,
            duration_ms=int((time.time() - started) * 1000),
        )

    async def _send_action_response(
        self,
        action_id: str,
        *,
        success: bool,
        result: Optional[dict[str, Any]] = None,
        error: Optional[str] = None,
        duration_ms: int = 0,
    ) -> None:
        payload = HUPActionResponsePayload(
            action_id=action_id,
            success=success,
            result=result or {},
            error=error,
            duration_ms=duration_ms,
        )
        await self._send("hup_action_response", payload)

    async def _heartbeat_loop(self) -> None:
        while self._ws is not None and not self._stop.is_set():
            try:
                await asyncio.sleep(self._heartbeat_ms / 1000.0)
                await self._send(
                    "node_heartbeat",
                    NodeHeartbeatPayload(ts=time.time()),
                )
            except Exception as exc:
                logger.debug("heartbeat stopped: %s", exc)
                return

    async def _send(self, type_: str, payload: Any) -> None:
        if self._ws is None:
            logger.debug("dropping %s frame — not connected", type_)
            return
        model_cls = MESSAGE_TYPES.get(type_)
        if model_cls is None:
            raise ValueError(f"unknown HUP type: {type_!r}")
        if not isinstance(payload, model_cls):
            payload = model_cls(**(payload or {}))
        frame = {
            "hup_version": HUP_VERSION,
            "type": type_,
            "ts": time.time(),
            "payload": payload.model_dump(exclude_none=False),
        }
        async with self._send_lock:
            await self._ws.send(json.dumps(frame))


__all__ = ["FeralNode"]
